lib/wallet/src/Cardano/Wallet/DB/Store/DeltaUTxO/Model.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Cardano.Wallet.DB.Store.DeltaUTxO.Model (
+    -- * Types
+    UTxOHistory,
+
+    -- * Accessors
+    getTip,
+    getFinality,
+    empty,
+    getUTxO,
+
+    -- * Changes
+    DeltaUTxOHistory (..),
+
+    -- * For testing
+    getSpent,
+) where
+
+import Prelude
+
+import Cardano.Slotting.Slot
+    ( SlotNo, WithOrigin (..) )
+import Cardano.Wallet.DB.Store.DeltaUTxO.Model.Internal
+    ( UTxOHistory (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Slot )
+import Cardano.Wallet.Primitive.Types.Tx.TxIn
+    ( TxIn )
+import Cardano.Wallet.Primitive.Types.UTxO
+    ( DeltaUTxO (..), UTxO, dom, excluding )
+import Data.Delta
+    ( Delta (..) )
+import Data.Foldable
+    ( fold )
+import Data.Map.Strict
+    ( Map )
+import Data.Set
+    ( Set )
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+-- | Changes to the UTxO history.
+data DeltaUTxOHistory
+    = -- | New slot tip, changes within that block.
+      AppendBlock SlotNo DeltaUTxO
+    | -- | Rollback tip.
+      Rollback Slot
+    | -- | Move finality forward.
+      Prune Slot
+
+-- | An empty UTxO history
+empty :: UTxO -> UTxOHistory
+empty utxo =
+    UTxOHistory
+        { history = utxo
+        , creationSlots = Map.singleton Origin $ dom utxo
+        , spentSlots = mempty
+        , tip = Origin
+        , finality = Nothing
+        }
+
+-- | Returns the UTxO.
+getUTxO :: UTxOHistory -> UTxO
+getUTxO UTxOHistory{history, spentSlots} = history `excluding` fold spentSlots
+
+-- | Returns the tip slot.
+getTip :: UTxOHistory -> Slot
+getTip UTxOHistory{tip} = tip
+
+-- | Returns the finality slot.
+getFinality :: UTxOHistory -> Maybe Slot
+getFinality UTxOHistory{finality} = finality
+
+revertMapOfSets :: Ord v => Map k (Set v) -> Map v k
+revertMapOfSets m = Map.fromList $ do
+    (k, vs) <- Map.toList m
+    v <- Set.toList vs
+    pure (v, k)
+
+-- | Returns the spent TxIns that can be rolled back.
+getSpent :: UTxOHistory -> Map TxIn SlotNo
+getSpent UTxOHistory{spentSlots} = revertMapOfSets spentSlots
+
+-- keep only non-empty slots
+cleanUpSlots :: Foldable t => Map s (t a) -> Map s (t a)
+cleanUpSlots = Map.filter (not . null)
+
+-- how to apply a DeltaUTxOHistory to a UTxOHistory
+instance Delta DeltaUTxOHistory where
+    type Base DeltaUTxOHistory = UTxOHistory
+    apply
+        (AppendBlock slot delta)
+        noop@UTxOHistory{history, spentSlots, creationSlots, tip, finality}
+            | At slot <= tip = noop
+            | otherwise =
+                UTxOHistory
+                    { history = history <> received delta
+                    , creationSlots =
+                        cleanUpSlots $
```

Comment:
Could you move `cleanUpSlots` "inside" the `Map.insert` operation, i.e. instead of inserting a potentially `mempty` value, restrict the application of `insert` only to those cases in which `dom (received delta) …` is empty? 🤔

Perhaps defining a helper function

```hs
insertNonMempty :: Monoid v => k -> v -> Map k v -> Map k v
insertNonMempty k v m = if v == mempty then m else Map.insert k v m
```

is the way to go.


I understand that `cleanUpSlots` is an "obviously correct" way of enforcing the invariant that all elements of the `Map` are `≠ mempty`. But unfortuately, it's expensive as it becomes a linear scan over the entire map, taking time O(n) where n is the size of `history`. At this point, we want `apply` to take time O(m * log n) where m is the "size" of `delta`.

Reply:
cleanUpSlots is technically O(n) of the slots but yeah if the sets are singletons it's O(n) of the utxo

Reply:
fixed